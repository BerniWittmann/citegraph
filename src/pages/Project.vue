<template>
  <div class="project">
    <v-alert v-if="!hasProject" type="error" prominent>
      {{ $t('project.not_available') }}
    </v-alert>

    <div v-if="hasProject">
      <h1>{{ project.name }}</h1>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project from '@/models/project'

@Component
export default class ProjectPage extends Vue {
  get project (): Project | undefined {
    return this.$store.getters['projects/activeProject']
  }

  get hasProject (): boolean {
    return this.$store.getters['projects/hasActiveProject']
  }
}
</script>
