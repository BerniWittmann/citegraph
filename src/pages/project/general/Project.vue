import { ProjectStates } from '@/models/project'
<template>
  <div class="project">
    <div class="d-flex mr-5 mt-4">
      <h1>{{ project.name }}</h1>
      <v-spacer/>
      <v-btn :to="{ name: 'projects.single.edit', params: { projectId: project.id } }">
        <v-icon>mdi-pencil</v-icon>
        {{ $t('project.edit.button_text') }}
      </v-btn>
    </div>
    <v-container v-if="hasNoDataYet">
      <v-row justify="center">
        <v-col :lg="3" class="d-flex justify-center flex-column text-center">
          <empty-icon class="align-self-center mb-6"></empty-icon>
          <h3 class="headline">{{ $t('project.single.no_data.description') }}</h3>
          <p>{{ $t('project.single.no_data.import_data') }}</p>
          <v-btn color="primary" rounded large :to="{ name: 'projects.single.import', params: { projectId: project.id } }">
            {{ $t('project.single.no_data.button') }}
          </v-btn>
        </v-col>
      </v-row>
    </v-container>
    <v-container v-else>
      <v-row v-if="datasetInformation">
        <v-col>
          <v-card>
            <v-card-title>{{ $t('project.single.dataset.title') }}</v-card-title>
            <v-card-text>
              <v-container>
                <v-row>
                  <v-col v-for="(count, name) in datasetInformation" :key="name">
                    <v-card outlined class="d-flex flex-column justify-center text-center pa-5">
                      <p class="headline">{{count}}</p>
                      <p class="mb-0">{{ $t('project.single.dataset.' + name ) }}</p>
                    </v-card>
                  </v-col>
                </v-row>
              </v-container>
            </v-card-text>
          </v-card>
        </v-col>
      </v-row>
      <v-row>
        <v-col lg="3" md="6" cols="12" v-if="project.topAuthors">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_authors') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="author in project.topAuthors" :key="author.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'author',
                 entityId: author.id
                }}">
                  {{ author.firstName }} {{ author.lastName }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.topRecords">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_records') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="record in project.topRecords" :key="record.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'record',
                 entityId: record.id
                }}">
                  {{ record.title }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.countryRelevanceData">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.country_relevance') }}
            </v-card-title>
            <v-card-text>
              TODO Diagram here
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12">
          <v-row no-gutters>
            <v-col v-if="project.topKeywords" class="mb-6">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.top_10_keywords') }}
                </v-card-title>
                <v-card-text>
                  <v-chip-group multiple column>
                    <v-chip v-for="keyword in project.topKeywords" :key="keyword">{{ keyword }}</v-chip>
                  </v-chip-group>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
          <v-row no-gutters>
            <v-col v-if="project.counts">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.visualizations') }}
                </v-card-title>
                <v-card-text>
                  <div class="d-flex flex-column justify-center text-center pa-5" v-if="hasNoVisualizationsYet">
                    <p class="mb-0">{{ $t('project.single.overview.no_visualizations_yet') }}</p>
                    <v-btn color="primary" rounded>{{ $t('project.single.overview.add_visualization') }}</v-btn>
                  </div>
                  <v-card outlined class="d-flex flex-column justify-center text-center pa-5" v-else>
                    <p class="headline">{{ project.counts.visualizations }}</p>
                    <p class="mb-0">{{ $t('project.single.overview.visualizations') }}</p>
                  </v-card>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project, { ProjectStates } from '@/models/project'
import EmptyIcon from '@/assets/svgs/Empty.vue'

@Component({
  components: {
    EmptyIcon
  }
})
export default class ProjectPage extends Vue {
  get project (): Project {
    return this.$store.getters['projects/activeProject']
  }

  get datasetInformation (): Object | undefined {
    if (!this.project.counts) return undefined
    const data = { ...this.project.counts }
    delete data['visualizations']
    return data
  }

  get hasNoDataYet (): boolean {
    return this.project.state === ProjectStates.IMPORT
  }

  get hasNoVisualizationsYet (): boolean {
    return this.project.state === ProjectStates.EXPLORE
  }
}
</script>
