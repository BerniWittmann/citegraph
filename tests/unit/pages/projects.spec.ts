import { shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import ProjectsPage from '@/pages/projects/Projects.vue'
import Project from '@/models/project'
import EmptySlotComponent from '../EmptySlotComponent.vue'

describe('pages/Projects.vue', () => {
  const projects = [
    new Project({ id: 1, name: 'First Project' }),
    new Project({ id: 2, name: 'Second Project' }),
    new Project({ id: 12, name: 'My awesome Project' }),
    new Project({ id: 42, name: 'Other Project' })
  ]
  const dispatch = jest.fn().mockReturnValue({
    then: (cb: Function) => cb()
  })
  const router = {
    push: jest.fn()
  }
  const getWrapper = (currentProjects: Array<Project>) => {
    return shallowMount(ProjectsPage, {
      i18n,
      stubs: {
        VTooltip: EmptySlotComponent
      },
      mocks: {
        $store: {
          getters: {
            'projects/projects': currentProjects
          },
          dispatch
        },
        $router: router
      }
    })
  }
  it('renders the page', () => {
    const wrapper = getWrapper(projects)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the projects', () => {
    const wrapper = getWrapper(projects)
    const projectCards = wrapper.findAll('.project-card')
    expect(projectCards.length).toEqual(4)
  })
  it('a click on a project card navigates to the project', () => {
    const wrapper = getWrapper(projects)
    const projectCards = wrapper.findAll('.project-card')
    expect(projectCards.length).toEqual(4)
    const openButton = projectCards.at(2).find('v-btn-stub')
    expect(openButton.exists()).toBeTruthy()

    // @ts-ignore
    wrapper.vm.openProject(projects[2])

    expect(dispatch).toHaveBeenCalledWith('projects/openProject', projects[2])
    expect(router.push).toHaveBeenCalledWith({ name: 'projects.single', params: { projectId: '12' } })
  })
  it('can handle if no projects are available', () => {
    const wrapper = getWrapper([])
    expect(wrapper.html()).toMatchSnapshot()
    const projectCards = wrapper.findAll('.project-card')
    expect(projectCards.length).toEqual(0)
  })
  it('renders a button to add projects', () => {
    const wrapper = getWrapper(projects)
    const addButton = wrapper.find('.add-card')
    expect(addButton.exists()).toBeTruthy()
    expect(addButton.props('to')).toEqual({ name: 'projects.add' })
  })

  it('a project can be deleted', () => {
    const wrapper = getWrapper(projects)
    // @ts-ignore
    wrapper.vm.deleteProject(projects[1])
    expect(dispatch).toHaveBeenCalledWith('projects/deleteProject', projects[1])
  })

  it('a project can be edited', () => {
    const wrapper = getWrapper(projects)
    const projectCard = wrapper.find('.project-card')
    const link = projectCard.find('v-list-item-stub')
    expect(link.exists()).toBeTruthy()
    expect(link.text()).toEqual('project.edit.button_text')
    expect(link.props('to')).toEqual({
      name: 'projects.single.edit', params: { projectId: 1 }
    })
  })
})
