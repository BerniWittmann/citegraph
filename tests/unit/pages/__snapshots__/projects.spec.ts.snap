// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Projects.vue can handle if no projects are available 1`] = `
<div class="projects">
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <h1 class="display-1 mt-5">projects.your_projects</h1>
    </v-row-stub>
    <v-row-stub tag="div">
      <v-col-stub cols="3" tag="div">
        <v-card-stub loaderheight="4" to="[object Object]" tag="div" height="100%" minheight="130" hover="true" outlined="true" class="d-flex justify-center align-center text--primary add-card">
          <v-icon-stub color="primary">mdi-plus</v-icon-stub>
        </v-card-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;

exports[`pages/Projects.vue renders the page 1`] = `
<div class="projects">
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <h1 class="display-1 mt-5">projects.your_projects</h1>
    </v-row-stub>
    <v-row-stub tag="div">
      <v-col-stub cols="3" tag="div">
        <v-card-stub loaderheight="4" tag="div" hover="true" class="project-card">
          <v-card-text-stub>
            <p class="title text--primary">
              First Project
            </p>
          </v-card-text-stub>
          <v-card-actions-stub>
            <v-btn-stub color="primary accent-4" tag="button" activeclass="" to="[object Object]" text="true" type="button">
              projects.open_project
            </v-btn-stub>
            <v-spacer-stub></v-spacer-stub>
            <v-menu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" offsety="true" origin="top left" transition="v-menu-transition">
              <v-list-stub tag="div" tile="true">
                <v-list-item-stub activeclass="" tag="div">
                  <v-list-item-title-stub class="error--text">project.delete</v-list-item-title-stub>
                </v-list-item-stub>
              </v-list-stub>
            </v-menu-stub>
          </v-card-actions-stub>
        </v-card-stub>
      </v-col-stub>
      <v-col-stub cols="3" tag="div">
        <v-card-stub loaderheight="4" tag="div" hover="true" class="project-card">
          <v-card-text-stub>
            <p class="title text--primary">
              Second Project
            </p>
          </v-card-text-stub>
          <v-card-actions-stub>
            <v-btn-stub color="primary accent-4" tag="button" activeclass="" to="[object Object]" text="true" type="button">
              projects.open_project
            </v-btn-stub>
            <v-spacer-stub></v-spacer-stub>
            <v-menu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" offsety="true" origin="top left" transition="v-menu-transition">
              <v-list-stub tag="div" tile="true">
                <v-list-item-stub activeclass="" tag="div">
                  <v-list-item-title-stub class="error--text">project.delete</v-list-item-title-stub>
                </v-list-item-stub>
              </v-list-stub>
            </v-menu-stub>
          </v-card-actions-stub>
        </v-card-stub>
      </v-col-stub>
      <v-col-stub cols="3" tag="div">
        <v-card-stub loaderheight="4" tag="div" hover="true" class="project-card">
          <v-card-text-stub>
            <p class="title text--primary">
              My awesome Project
            </p>
          </v-card-text-stub>
          <v-card-actions-stub>
            <v-btn-stub color="primary accent-4" tag="button" activeclass="" to="[object Object]" text="true" type="button">
              projects.open_project
            </v-btn-stub>
            <v-spacer-stub></v-spacer-stub>
            <v-menu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" offsety="true" origin="top left" transition="v-menu-transition">
              <v-list-stub tag="div" tile="true">
                <v-list-item-stub activeclass="" tag="div">
                  <v-list-item-title-stub class="error--text">project.delete</v-list-item-title-stub>
                </v-list-item-stub>
              </v-list-stub>
            </v-menu-stub>
          </v-card-actions-stub>
        </v-card-stub>
      </v-col-stub>
      <v-col-stub cols="3" tag="div">
        <v-card-stub loaderheight="4" tag="div" hover="true" class="project-card">
          <v-card-text-stub>
            <p class="title text--primary">
              Other Project
            </p>
          </v-card-text-stub>
          <v-card-actions-stub>
            <v-btn-stub color="primary accent-4" tag="button" activeclass="" to="[object Object]" text="true" type="button">
              projects.open_project
            </v-btn-stub>
            <v-spacer-stub></v-spacer-stub>
            <v-menu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" offsety="true" origin="top left" transition="v-menu-transition">
              <v-list-stub tag="div" tile="true">
                <v-list-item-stub activeclass="" tag="div">
                  <v-list-item-title-stub class="error--text">project.delete</v-list-item-title-stub>
                </v-list-item-stub>
              </v-list-stub>
            </v-menu-stub>
          </v-card-actions-stub>
        </v-card-stub>
      </v-col-stub>
      <v-col-stub cols="3" tag="div">
        <v-card-stub loaderheight="4" to="[object Object]" tag="div" height="100%" minheight="130" hover="true" outlined="true" class="d-flex justify-center align-center text--primary add-card">
          <v-icon-stub color="primary">mdi-plus</v-icon-stub>
        </v-card-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;
